apps/passport-client/src/dispatch.ts

```diff
@@ -463,20 +480,35 @@ async function sync(state: AppState, update: ZuUpdate) {
     return;
   }
 
+  // If we haven't downloaded from storage, do that first
   if (!state.downloadedPCDs && !state.downloadingPCDs) {
     console.log("[SYNC] sync action: download");
     update({
       downloadingPCDs: true
     });
 
-    const pcds = await downloadStorage();
+    /** 
+     * Get both PCDs and subscriptions
+     * Subscriptions might be null even if we got PCDs, if we were
+     * downloading from a pre-v3 version of encrypted storage 
+     * {@link SyncedEncryptedStorageV3}
+     * */
+    const { pcds, subscriptions } = await downloadStorage();
 
     if (pcds != null) {
       update({
         downloadedPCDs: true,
         downloadingPCDs: false,
         pcds: pcds,
-        uploadedUploadId: await pcds.getHash()
+        // If we got subscriptions, add them and generate an upload ID.
+        // The upload ID is a combination of hashes of the states of the PCDs
+        // and subscriptions.
+        // If later subscription-polling doesn't change the PCD set, we can
+        // avoid having to do an upload.
+        // If we didn't get subscriptions, it's because we downloaded from a
+        // pre-v3 version of encrypted storage and therefore we definitely
+        // want to do an upload.
+        ...(subscriptions !== null ? { uploadedUploadId: await makeUploadId(pcds, subscriptions), subscriptions } : {})
```

Comment:
This is a bit confusing.  Maybe it would be cleaner if the makeUploadId function simply knew how to generate hashes for null arguments (or make an empty FeedSubscriptionManager as standin)?  So long as it results in a hash different from the hash of a non-null value it should have the intended effect.

It think that would avoid the trickiness of the logic here.  In particular, the fact that you're expecting to forcing an upload by not changing the uploadedId feels fragile.  Is it guaranteed that the old value will trigger an upload?  I have trouble reasoning through that.  If you instead always generate a new value, and you're sure that the value of uploadedId without feed is different from the value with feeds, then I think the behavior is more obviously correct.

Reply:
The way I implemented this was to preserve the state machine logic as I found it, with the addition being that the hash of the subscription state is now an input into `uploadedId`.

If we had more (any!) unit tests around this state machine, I would feel a lot more comfortable refactoring it, but I think this should wait for a more comprehensive review. I like the approach laid out in your notion doc, and I think we can address these concerns when we implement that.

Reply:
I agree with your sentiment of not wanting to change the sync logic, and also wish there were more tests.  I think my suggestion here isn't to change the syncing logic, but effectively to move the null check inside of the makeUploadedId function.

Not a huge deal either way.  I'll likely be rewriting this logic soon, and will try to give it more tests in the process.